File: PlatformService/src/main/java/com/cognizant/devops/platformservice/agentmanagement/service/AgentManagementService.java
/*******************************************************************************
 * Copyright 2017 Cognizant Technology Solutions
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 ******************************************************************************/

package com.cognizant.devops.platformservice.agentmanagement.service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.cognizant.devops.platformcommons.exception.InsightsCustomException;

public interface AgentManagementService {

	public String registerAgent(String toolName, String agentVersion, String osversion, String configDetails,
			String trackingDetails, boolean vault) throws InsightsCustomException;

	public String uninstallAgent(String agentId, String toolName, String osversion) throws InsightsCustomException;

	public String startStopAgent(String agentId, String toolName, String osversion, String action)
			throws InsightsCustomException;

	public String updateAgent(String agentId, String configDetails, String toolName, String agentVersion,
			String osversion, boolean vault) throws InsightsCustomException;

	// This is used during Agent registration, provide list of Agents, with version
	// for offline processing
	public Map<String, ArrayList<String>> getOfflineSystemAvailableAgentList() throws InsightsCustomException;

	// This is used during Agent registration, provide list of Agents, with version
	// for online processing,for docroot
	public Map<String, ArrayList<String>> getDocrootAvailableAgentList() throws InsightsCustomException;

	// This is used during Agent registration, provide list of Agents, with version
	// for online processing,for nexus
	public Map<String, ArrayList<String>> getRepoAvailableAgentList() throws InsightsCustomException;

	// For agent registration, gives you RAW config.json from docroot
	public String getToolRawConfigFile(String version, String tool) throws InsightsCustomException;

	// Provides currently registered Agents in DB
	public List<AgentConfigTO> getRegisteredAgents() throws InsightsCustomException;

	// Returns config.json for request AgentId from Db
	public AgentConfigTO getAgentDetails(String agentId) throws InsightsCustomException;
}

File: PlatformUI3/src/app/com/cognizant/devops/platformui/modules/admin/agent-management/agent-management-service.ts
/*******************************************************************************
 * Copyright 2019 Cognizant Technology Solutions
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 ******************************************************************************/
import { Injectable } from '@angular/core';
import { RestCallHandlerService } from '@insights/common/rest-call-handler.service';
import { Observable, throwError } from 'rxjs';

export interface IAgentService {
    loadGlobalHealthConfigurations(): Promise<any>;
    loadHealthConfigurations(toolName: string, toolCategory: string): Promise<any>;
    loadServerHealthConfiguration(ServerName: string): Promise<any>;
    getDocRootAgentVersionTools(): Promise<any>;
    getDocrootAgentConfig(Version: string, toolName: string): Promise<any>;
    getDbAgentConfig(agentId: string): Promise<any>;
    loadAgentServices(ServerName: string): Promise<any>;
    registerAgent(toolName: string, toolVersion: string, osName: string, configData: string, trackingDetails: string): Promise<any>;
    updateAgent(agentId: string, configData: string, toolName: string, toolVersion: string, osName: string): Promise<any>;
    agentStartStop(agentId: string, toolName: string, osName: string, actionType: string): Promise<any>;
    agentUninstall(agentId: string, toolName: string, osversion: string): Promise<any>
}

@Injectable()
export class AgentService implements IAgentService {

    constructor(private restCallHandlerService: RestCallHandlerService) {

    }

    loadGlobalHealthConfigurations(): Promise<any> {
        return this.restCallHandlerService.get("HEALTH_GLOBAL");
    }

    loadHealthConfigurations(toolName: string, toolCategory: string): Promise<any> {
        return this.restCallHandlerService.get("HEALTH_TOOL", { 'tool': toolName, 'category': toolCategory });
    }

    loadServerHealthConfiguration(ServerName: string): Promise<any> {
        return this.restCallHandlerService.get(ServerName);
    }

    getDocRootAgentVersionTools(): Promise<any> {
        return this.restCallHandlerService.get("DOCROOT_AGENT_VERSION_TOOLS");
    }

    getDocrootAgentConfig(Version: string, toolName: string): Promise<any> {
        return this.restCallHandlerService.get("DOCROOT_AGENT_TOOL_CONFIG_DETAILS", { 'version': Version, 'tool': toolName });
    }

    getDbAgentConfig(agentId: string): Promise<any> {
        return this.restCallHandlerService.get("DB_AGENT_CONFIG_DETAILS", { 'agentId': agentId });
    }

    loadAgentServices(ServerName: string): Promise<any> {
        return this.restCallHandlerService.get(ServerName);
    }

    registerAgent(toolName: string, toolVersion: string, osName: string, configData: string, trackingDetails: string): Promise<any> {
        return this.restCallHandlerService.postWithParameter("AGENT_REGISTER", { 'toolName': toolName, 'agentVersion': toolVersion, 'osversion': osName, 'configDetails': configData, 'trackingDetails': trackingDetails }, { 'Content-Type': 'application/x-www-form-urlencoded' }).toPromise();
    }

    updateAgent(agentId: string, configData: string, toolName: string, toolVersion: string, osName: string): Promise<any> {
        return this.restCallHandlerService.postWithParameter("AGENT_UPDATE", { 'agentId': agentId, 'configJson': configData, 'toolName': toolName, 'agentVersion': toolVersion, 'osversion': osName }, { 'Content-Type': 'application/x-www-form-urlencoded' }).toPromise();
    }

    agentStartStop(agentId: string, toolName: string, osName: string, actionType: string): Promise<any> {
        return this.restCallHandlerService.postWithParameter("AGENT_START_STOP", { 'agentId': agentId, 'toolName': toolName, 'osversion': osName, 'action': actionType }, { 'Content-Type': 'application/x-www-form-urlencoded' }).toPromise();
    }

    agentUninstall(agentId: string, toolName: string, osversion: string): Promise<any> {
        return this.restCallHandlerService.postWithParameter("AGENT_UNINSTALL", { 'agentId': agentId, 'toolName': toolName, 'osversion': osversion }, { 'Content-Type': 'application/x-www-form-urlencoded' }).toPromise();
    }

    registerAgentV2(registerAgentJson: string): Promise<any> {
        return this.restCallHandlerService.postWithAgentData("AGENT_REGISTERV2", registerAgentJson, "", { 'Content-Type': 'application/x-www-form-urlencoded' }).toPromise();
    }

    updateAgentV2(agentMappingJson: string): Promise<any> {
        return this.restCallHandlerService.postWithAgentData("AGENT_UPDATEV2", agentMappingJson, "", { 'Content-Type': 'application/x-www-form-urlencoded' }).toPromise();;
    }

}